engine/access/ingestion2/forest_manager.go

```diff
@@ -0,0 +1,369 @@
+func (fm *ForestManager) WorkerLoop(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	// periodic check to ensure we continue processing results even if no blocks are finalized
+	periodicCheck := time.NewTicker(forestManagerCheckInterval)
+	defer periodicCheck.Stop()
+
+	var lastChecked time.Time
+	notifierChan := fm.notifier.Channel()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-notifierChan:
+		case <-periodicCheck.C:
+			// skip if we already checked recently
+			// Note: this check will always be skipped on the happy path since block finalization
+			// should happen more frequently than the check interval.
+			if time.Since(lastChecked) < forestManagerCheckInterval {
+				continue
+			}
+		}
+		lastChecked = time.Now()
+
+		// TODO: we may want to only run the full check after dropping below some threshold to reduce
+		// the overhead of a full traversal when only a few results can be added to the queue.
+		if uint(fm.workQueue.Len()) >= fm.maxQueueSize {
+			continue
+		}
+
+		latestPersistedSealedResultContainer, err := fm.getLatestPersistedSealedResultContainer()
+		if err != nil {
+			ctx.Throw(err)
+			return
+		}
+
+		fm.traverseDescendants(latestPersistedSealedResultContainer)
+	}
+}
+
+// OnReceiptAdded is called when a new receipt is added to the forest.
+// This is used to notify the manager that a result and its descendants may be processable.
+func (fm *ForestManager) OnReceiptAdded(container *ExecutionResultContainer) {
+	if container.IsEnqueued() {
+		return // already enqueued, adding this receipt will not change processability
+	}
+
+	if uint(fm.workQueue.Len()) >= fm.maxQueueSize {
+		return
+	}
+
+	parentID, _ := container.Parent()
+	parent, found := fm.resultsForest.GetContainer(parentID)
+	if !found {
+		return // parent does not exist in the forest yet, this result is not processable
+	}
+
+	// result previously was not enqueued and is now processable. this means that either this receipt
+	// caused the result to become processable, or there is more room in the queue.
+	// enqueue the result and check its descendants
+	if fm.checkAndEnqueueContainer(container, parent) {
+		fm.traverseDescendants(container)
+	}
+}
+
+// OnBlockStatusUpdated is called when a new block is finalized.
+// This is used to notify the forest that some results may now be processable.
+func (fm *ForestManager) OnBlockStatusUpdated(finalized, sealed *flow.Header) {
+	fm.mu.Lock()
+	defer fm.mu.Unlock()
+
+	fm.sealedHeight = sealed.Height
+	fm.finalizedHeight = finalized.Height
+
+	fm.notifier.Notify()
+}
+
+// getLatestPersistedSealedResultContainer returns the latest persisted sealed result's container.
+//
+// No errors are expected during normal operation.
+func (fm *ForestManager) getLatestPersistedSealedResultContainer() (*ExecutionResultContainer, error) {
+	latestPersistedSealedResultID, _ := fm.resultsForest.latestPersistedSealedResult.Latest()
```

Comment:
This is a similar instance:
* don't try to work of a global state (the [`LatestPersistedSealedResultReader`](https://github.com/onflow/flow-go/blob/b3f649ed221bea867d546ca176b7a3350dd6f6b9/engine/access/ingestion2/results_forest.go#L190) that underpins the forest)
* instead have the forest expose the information 


**Here is my recommendation in a nutshell:**

Just pretend there is no state you can query other than the Forest. The Forest mediates all your state access and can guarantees atomicity where needed (probably not very many places). The Forest can decide whether it returns its own knowledge, or queries the global state and returns that an its answer. Either way, it's the Forest that answers and decided what information it gives you.